lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+	} else {
+		// Check the BalanceEntry of the user.
+		transactorBalanceEntry, _, _ := bav.GetBalanceEntryForHODLerPubKeyAndCreatorPubKey(
+			txn.PublicKey,
+			txMeta.ProfilePublicKey.ToBytes(),
+			true)
+		if transactorBalanceEntry == nil || transactorBalanceEntry.isDeleted {
+			return 0, 0, nil,
+				errors.Wrapf(RuleErrorCoinLockupBalanceEntryDoesNotExist, "_connectCoinLockup")
+		}
+
+		// Validate the balance entry as having sufficient funds.
+		transactorBalanceNanos256 = transactorBalanceEntry.BalanceNanos.Clone()
+		if txMeta.LockupAmountBaseUnits.Gt(transactorBalanceNanos256) {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupInsufficientCoins, "_connectCoinLockup")
+		}
+
+		// We store the previous transactor balance entry in the event we need to revert the transaction.
+		prevTransactorBalanceEntry = transactorBalanceEntry
+
+		// Spend the transactor's DAO coin balance.
+		transactorBalanceEntry.BalanceNanos =
+			*uint256.NewInt().Sub(&transactorBalanceEntry.BalanceNanos, txMeta.LockupAmountBaseUnits)
+		bav._setDAOCoinBalanceEntryMappings(transactorBalanceEntry)
+	}
+
+	// By now we know the transaction to be valid. We now source yield information from either
+	// the profile's yield curve or the raw DeSo yield curve. Because there's some choice in how
+	// to determine the yield when the lockup duration falls between two profile specified yield curve
+	// points, we return here the two local points and choose/interpolate between them below.
+	leftYieldCurvePoint, rightYieldCurvePoint := bav.GetLocalYieldCurvePoints(profilePKID, lockupDurationNanoSeconds)
+
+	// Here we interpolate (choose) the yield between the two returned local yield curve points.
+	//
+	// If we fall between two points, we choose the left yield curve point (i.e. the one with lesser lockup duration).
+	// The transactor earns yield only for the lockup duration specified by the left yield curve point but will
+	// be unable to unlock the coins until the transaction specified lockup duration expires.
+	txnYieldBasisPoints := uint64(0)
+	txnYieldEarningDurationNanoSecs := int64(0)
+	if leftYieldCurvePoint.LockupDurationNanoSecs < lockupDurationNanoSeconds {
+		txnYieldBasisPoints = leftYieldCurvePoint.LockupYieldAPYBasisPoints
+		txnYieldEarningDurationNanoSecs = leftYieldCurvePoint.LockupDurationNanoSecs
+	}
+	if rightYieldCurvePoint.LockupDurationNanoSecs == lockupDurationNanoSeconds {
+		txnYieldBasisPoints = rightYieldCurvePoint.LockupYieldAPYBasisPoints
+		txnYieldEarningDurationNanoSecs = rightYieldCurvePoint.LockupDurationNanoSecs
+	}
+
+	// Convert variables to a consistent uint256 representation. This is to use them in SafeUint256 math.
+	txnYieldBasisPoints256 := uint256.NewInt().SetUint64(txnYieldBasisPoints)
+	txnYieldEarningDurationNanoSecs256 := uint256.NewInt().SetUint64(uint64(txnYieldEarningDurationNanoSecs))
+
+	// Compute the yield associated with this operation, checking to ensure there's no overflow.
+	yieldFromTxn, err :=
```

Comment:
Something a little annoying that we may want to put a little juice into noodling on is how we can allow people to both stake their DESO *and* to leverage this lockup functionality.

For example, staking your DESO earns you yield based on how much we want to pay validators - let's call this the validator yield. But defining yield with these functions would enable a *secondary* yield that you can actually only access if you *don't* stake your DESO. Let's call this secondary yield the lockup yield.

I kindof wish we could find a clever way to combine these. Maybe you have to FIRST stake your DESO with these functions, and *then* you can allocate that LOCKED deso to PoS. This would allow you to earn BOTH the PoS yield and the P yield. But would require coordination with Sofonias and stuff.

I think what we can do is add this to our list of shit for PoS as a "nice to have." If we don't get to it, which we probably won't, then we just ship the "yield unification" change later on after PoS has fully stabilized. It will likely manifest as a secondary way to stake to PoS with locked DESO.